packages/expo-barcode-scanner/src/BarCodeScanner.tsx

```diff
@@ -0,0 +1,151 @@
+import { UnavailabilityError } from 'expo-errors';
+import mapValues from 'lodash.mapvalues';
+import PropTypes from 'prop-types';
+import React from 'react';
+import { findNodeHandle, Platform, ViewProps, ViewPropTypes } from 'react-native';
+
+import ExpoBarCodeScannerModule from './ExpoBarCodeScannerModule';
+import ExpoBarCodeScannerView from './ExpoBarCodeScannerView';
+
+const { BarCodeType, Type } = ExpoBarCodeScannerModule;
+
+const EVENT_THROTTLE_MS = 500;
+
+type BarCodeEvent = {
+  type: string;
+  data: string;
+  [key: string]: any;
+};
+
+export type BarCodeEventCallbackArguments = {
+  nativeEvent: BarCodeEvent;
+};
+
+export type BarCodeReadCallback = (params: BarCodeEvent) => void;
+
+export type BarCodeScannedCallback = (params: BarCodeEventCallbackArguments) => void;
+
+export interface BarCodeScannerProps extends ViewProps {
+  type?: 'front' | 'back' | number;
+  torchMode?: 'on' | 'off';
+  barCodeTypes?: string[];
+  onBarCodeRead?: BarCodeReadCallback;
+  onBarCodeScanned: BarCodeScannedCallback;
+}
+
+type AnyComponent = null | React.Component<any, any> | React.ComponentClass<any>;
+
+export default class BarCodeScanner extends React.Component<BarCodeScannerProps> {
+  lastEvents: { [key: string]: any };
+  lastEventsTimes: { [key: string]: any };
+  barCodeScannerRef: AnyComponent = null;
+  barCodeScannerHandle: null | number = null;
+
+  static Constants = {
+    BarCodeType,
+    Type,
+  };
+
+  static ConversionTables = {
+    type: Type,
+  };
+
+  static propTypes = {
+    ...ViewPropTypes,
+    onBarCodeScanned: PropTypes.func,
+    barCodeTypes: PropTypes.array,
+    type: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
+  };
+
+  static defaultProps = {
+    type: Type.back,
+    barCodeTypes: Object.values(BarCodeType),
+  };
+
+  constructor(props: BarCodeScannerProps) {
+    super(props);
+    this.lastEvents = {};
+    this.lastEventsTimes = {};
+  }
+
+  static async scanFromURLAsync(
+    url: string,
+    barCodeTypes: string[]
+  ): Promise<{ type: string; data: string }> {
+    if (!ExpoBarCodeScannerModule.scanFromURLAsync) {
+      throw new UnavailabilityError('expo-barcode-scanner', 'scanFromURLAsync');
+    }
+    if (Array.isArray(barCodeTypes) && !barCodeTypes.length) {
+      throw new Error('No barCodeTypes requested, provide at least one barCodeType for scanner');
+    }
+
+    if (Platform.OS === 'ios') {
+      if (Array.isArray(barCodeTypes) && !barCodeTypes.includes(BarCodeType.qr)) {
+        // Only QR type is supported on iOS, fail if one tries to use other types
+        throw new Error('Only QR type is supported by scanFromURLAsync() on iOS');
+      }
+      // on iOS use only supported QR type
+      return await ExpoBarCodeScannerModule.scanFromURLAsync(url, [BarCodeType.qr]);
+    }
+
+    // on Android if barCodeTypes not provided use all available types
+    const effectiveBarCodeTypes = barCodeTypes || Object.values(BarCodeType);
```

Comment:
Might be simpler to make the method parameter default to all the values up top: `barCodeTypes: string[] = Object.values(BarCodeType)` - what do you think?

Reply:
I can't see a reason not to have it.